### PR TITLE
Feature/metric calculation

### PR DESCRIPTION
**Pull request by ma4ypic4y**
Implemented KMeans clustering of embeddings.
Added computation of internal and external metrics:

- Internal metrics like silhouette score, Davies-Bouldin index, and Calinski-Harabasz index provide insights on how well the data points fit within their assigned clusters.

- External metrics such as accuracy, precision, recall, and F1 score offer an understanding of how accurately the clustering reflects the actual classifications.

### CODE REVIEW

File: src/training.py [hunk 1]
```diff
@@ -87,17 +104,122 @@ def training_step(self, batch, batch_idx):
...
+        labels : np.ndarray
+            The actual labels for the given embeddings.
+
+        Returns
+        -------
+        np.ndarray
+            The labels predicted by KMeans mapped according to maximum intersection with actual labels.
+        """
+
+        kmeans = KMeans(n_clusters=40, random_state=0, n_init=10).fit(embeddings)
```
Comment (Kaliostrogoblin) [inline src/training.py]: make please `n_clusters = np.unique(labels).size` or create a separate param with explicit defining of clusters number

File: src/metrics.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+class BaseScoreMetric(Metric):
+    def __init__(self, score_func, **kwargs):
+        super().__init__()
+        self.score = 0
+        self.score_func = score_func
+        self.score_func_kwargs = kwargs
+
+    def update(self, *args, **kwargs):
+        self.score = self.score_func(*args, **kwargs, **self.score_func_kwargs)
```
Comment (Kaliostrogoblin) [inline src/metrics.py]: As I understood, you are saving only the latest metric. It is okay for training epoch, but doesn't work well with evaluation (validation) step. I propose to either append every score to the list of scores and in the `compute` return the mean and clear the list, or you could have a total sum of metrics and number of updates, so in the end you can just divide total sum by number of updates. Of course, you should clear both sums in the compute before returning result

Comment (Kaliostrogoblin) [reply src/metrics.py]: For the recall, precision and F1 you just need to create external configurables from them for gin and import them somewhere. See the example here - https://github.com/t3hseus/ariadne/blob/master/ariadne/external_configurables.py

File: src/training.py [hunk 3]
```diff
@@ -1,21 +1,75 @@
...
 from .visualization import draw_embeddings
 
 
+@gin.configurable(allowlist=None)
+def clustering_algorithm(module: str, class_: str, **kwargs: dict):
+    module = importlib.import_module(module)
+    cls = getattr(module, class_)
+    return cls(**kwargs)
+
+class Clustering:
```
Comment (Kaliostrogoblin) [inline src/training.py]: I am proposing to move this code to a separate clustering.py file in src

File: src/training.py [hunk 4]
```diff
@@ -93,15 +157,30 @@ def training_step(self, batch, batch_idx):
...
         loss, embeddings = self._forward_batch(batch, return_embeddings=True)
-        # save predictions for visualization
+        emb_np = embeddings.cpu().detach().numpy()
+        evt_ids_np = batch[1].cpu().numpy()
+        cluster_preds = self.clustering.cluster_and_link(emb_np, evt_ids_np)
+
+        metrics_results = {}
+
+        for metric in self.metrics:
+            if isinstance(metric, (F1ScoreMetric, PrecisionScoreMetric, RecallScoreMetric, AccuracyScoreMetric)):
```
Comment (Kaliostrogoblin) [inline src/training.py]: reformat long string please to make it shorter, you can spread it over multiple lines

File: src/training.py [hunk 5]
```diff
@@ -93,15 +157,30 @@ def training_step(self, batch, batch_idx):
...
+                metric.update(evt_ids_np, cluster_preds)
+            elif isinstance(metric, BaseScoreMetric):
+                metric.update(emb_np, cluster_preds)
+
+            metric_name = type(metric).__name__
+            self.epoch_metrics[metric_name].append(metric.compute())
```
Comment (Kaliostrogoblin) [inline src/training.py]: not sure that this should work in a such manner. Better to avoid this boilerplate code by using a proper realization of compute and update. E.g.
image

Comment (Kaliostrogoblin) [reply src/training.py]: https://torchmetrics.readthedocs.io/en/stable/pages/lightning.html

File: configs/train.cfg [hunk 6]
```diff
@@ -14,12 +14,29 @@ experiment.triplet_margin = 0.1
...
+    @RecallScoreMetric(a)
+]
+#experiment.resume_from_checkpoint =
 
 ### model ###
 TrackEmbedder.n_blocks = 2
 TrackEmbedder.output_dim = 32
 
+### clustering model ###
+clustering_algorithm.module = 'sklearn.cluster'
```
Comment (Kaliostrogoblin) [inline configs/train.cfg:34]: I am not sure, but maybe it is better to pass it explicitly to experiment arguments
